### PR TITLE
feat(event-source): add Alexa Smart Home as an event source

### PR DESCRIPTION
**Pull request by sanathkr**
*Issue #, if available:*

*Description of changes:*
Live pull request from Twitch. @Twitch Audience: Please review my code and give me feedback.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by satblip** [comment]
Live pull request, nice :-)

**Comment by bvanneerven** [comment]
This misses unit tests IMHO ;) ;)

File: tests/translator/input/alexa_smart_home.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# File: sam.yml
+# Version: 0.9
```
Comment (brettstack) [inline tests/translator/input/alexa_smart_home.yaml]: Remove?

File: samtranslator/model/eventsources/push.py [hunk 2]
```diff
@@ -484,6 +484,23 @@ def to_cloudformation(self, **kwargs):
         return resources
 
 
+class AlexaSmartHome(PushEventSource):
+    resource_type = 'AlexaSmartHome'
+    principal = 'alexa-connectedhome.amazon.com'
+
+    property_types = {}
```
Comment (jfuss) [inline samtranslator/model/eventsources/push.py]: Shouldn't this have an 'EventSourceToken' Property?

File: samtranslator/model/eventsources/push.py [hunk 3]
```diff
@@ -484,6 +484,23 @@ def to_cloudformation(self, **kwargs):
         return resources
 
 
+class AlexaSmartHome(PushEventSource):
```
Comment (jfuss) [inline samtranslator/model/eventsources/push.py]: We are lacking pretty hard in the unit tests. I love the functional tests but we can do better.

This file is also getting fairly large. Can we split this out into its own class? What are your thoughts around that?

Comment (brettstack) [inline samtranslator/model/eventsources/push.py]: Move this to `alexa-smart-home.py` (or `alexasmarthome.py` - need to decided on naming convention)

Comment (brettstack) [reply samtranslator/model/eventsources/push.py]: We did this for CloudWatchLogs event and should continue with this approach.

**Comment by brettstack** [comment]
Missing example, documentation, and spec update. I don't think we need to tackle unit tests for event sources in this PR but we should get that going soon #364

**Comment by jfuss** [comment]
I think we also need the validator schema updated too right?

We also didn’t add any tests for intrinsic functions. Are we supporting all or a subset and why?

File: samtranslator/model/eventsources/push.py (hunk 2 above)
Comment (brettstack) [reply samtranslator/model/eventsources/push.py]: Can you update the existing Alexa Skill with this also?

**Comment by ScriptAutomate** [comment]
Any updates on this, as it has been >8 months since the last comment update on here? This is a blocker of mine, as detailed in #613, while I use the following workaround after a `sam-cli deploy`:

* [[Gist] Adding an Alexa Skills Kit Permission w/ InvokeFunction Permissions After Deploying w/ SAM CLI](https://gist.github.com/ScriptAutomate/54ab687ff560e85191db1ff8c732e7f8)

**Comment by ScriptAutomate** [comment]
Additional Comment:

After testing today, it looks like CFN already may support this functionality, though it can only currently be done outside of the serverless transform via standard CFN resources.

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Description: Alexa App Hello World Lambda Endpoint

Mappings:
    Variables:
        AlexaSkillKit:
            Id: amzn1.ask.skill.xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx

Globals:
    Function:
        Timeout: 3

Resources:

    HelloWorldFunction:
        Type: AWS::Serverless::Function
        Properties:
            CodeUri: hello_world/
            Handler: app.lambda_handler
            Runtime: python3.6

    HelloWorldFunctionAskPermission:
        Type: AWS::Lambda::Permission
        DependsOn: HelloWorldFunction
        Properties:
            Action: lambda:InvokeFunction
            EventSourceToken: !FindInMap
                - Variables
                - AlexaSkillKit
                - Id
            FunctionName: !GetAtt HelloWorldFunction.Arn
            Principal: alexa-appkit.amazon.com

Outputs:

    HelloWorldFunction:
        Description: "Alexa Hello World Lambda Function ARN"
        Value: !GetAtt HelloWorldFunction.Arn

    HelloWorldFunctionIamRole:
        Description: "Implicit IAM Role created for Alexa Hello World function"
        Value: !GetAtt HelloWorldFunctionRole.Arn

    HelloWorldFunctionAlexaSkillKitId:
        Description: "Alexa Skill Permitted Lambda Invokation Permissions"
        Value: !FindInMap
            - Variables
            - AlexaSkillKit
            - Id
```

I haven't created Alexa Smart Home permissions before, but it looks like the principal would be `alexa-connectedhome.amazon.com` instead of `alexa-appkit.amazon.com`.

So, the above should work in `template.yaml` configurations with `sam-cli` without need of added capability in the serverless transform (for those who have wanted to figure out how to do this without a transform shorthand support).

**Comment by jlhood** [comment]
@kienpham2000 Thanks for reaching out. This PR hasn't been worked on since it was initially open, but we'd love to have this feature in SAM. If you want to take it on, you can open your own PR against the `develop` branch using this PR as an example.

If you're new to contributing to SAM, [this](https://github.com/awslabs/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.rst) is a good place to start.

**Comment by ragurney** [comment]
Thanks @ScriptAutomate that worked for me! Also @sanathkr was that live stream recorded by any chance? I would love to see your process for contributing to this repo.

**Comment by ShreyaGangishetty** [comment]
@ragurney Please checkout the [this link](https://www.youtube.com/watch?v=uxv1dOExq5U&feature=youtu.be)

**Comment by jfuss** [comment]
Closing this as this has become inactive for quite some time. If this feature is useful to you, I would encourage you to submit a new PR with the changes.

**Pull request closed without merging**